pkg/net/url.go

```diff
@@ -0,0 +1,102 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net
+
+import (
+	"encoding/json"
+	"errors"
+	"net/url"
+	"path"
+)
+
+// URL - improved JSON friendly url.URL.
+type URL url.URL
+
+// IsEmpty - checks URL is empty or not.
+func (u URL) IsEmpty() bool {
+	return u.String() == ""
+}
+
+// String - returns string representation of URL.
+func (u URL) String() string {
+	// if port number 80 and 443, remove for http and https scheme respectively
+	if u.Host != "" {
+		host := MustParseHost(u.Host)
+		switch {
+		case u.Scheme == "http" && host.Port == 80:
+			fallthrough
+		case u.Scheme == "https" && host.Port == 443:
+			u.Host = host.Name
+		}
+	}
+
+	uu := url.URL(u)
+	return uu.String()
+}
+
+// MarshalJSON - converts to JSON string data.
+func (u URL) MarshalJSON() ([]byte, error) {
+	return json.Marshal(u.String())
+}
+
+// UnmarshalJSON - parses given data into URL.
+func (u *URL) UnmarshalJSON(data []byte) (err error) {
+	var s string
+	if err = json.Unmarshal(data, &s); err != nil {
+		return err
+	}
+
+	// Allow empty string
+	if s == "" {
+		return nil
```

Comment:
Here also, we would need to clear `u` to remove any stale value.

Reply:
Done